### ISSUE TITLE
libssp-0.dll broken in mingw-w64-x86_64-gcc-libs-12.2.0-5 package

### ISSUE THREAD

**Issue by nga888** [msys2/MINGW-packages#13830]
### Description

After updating to mingw-w64-x86_64-gcc-libs-12.2.0-5, I get the following error when running `git --version`:

`*** stack smashing detected ***:  terminated`

Using `git` from the package git-for-windows/mingw-w64-x86_64-git 2.38.1.1.b85c8f604d-1.

Switching to `libssp-0.dll` from the previous package appears to work fine.

### Verification

- [X] I have verified that my MSYS2 is up-to-date before submitting the report (see https://www.msys2.org/docs/updating/)

### Windows Version

Microsoft Windows [Version 10.0.19044.2130]

### MINGW environments affected

- [X] MINGW64
- [ ] MINGW32
- [ ] UCRT64
- [ ] CLANG64
- [ ] CLANG32
- [ ] CLANGARM64

### Expected behavior

`git` to run without terminating.

### Actual behavior

`git` terminates with the error: `*** stack smashing detected ***:  terminated`.

### Repro steps

Run `git --version` with `git` from the package git-for-windows/mingw-w64-x86_64-git 2.38.1.1.b85c8f604d-1.

**Comment by nga888** [comment]
I did try building the mingw-w64-x86_64-gcc-libs-12.2.0-5 package myself and I noticed there were some warnings about infinite recursion in the build of `libssp-0.dll`. Could this be the cause?

**Comment by revelator** [comment]
was the libssp package not replaced by the crt fortification improvement ? might have been a snag if it was still enabled in your build mine has it disabled after updating the packages from git.

**Comment by nga888** [comment]
Yes, `libssp` has been replaced/deprecated but the `git` package that I'm using from the GFW repo (https://wingit.blob.core.windows.net/x86-64) is still using `libssp`. There's probably not much else that still relies on `libssp-0.dll` but this build of `git` still does.

In any case, the `libssp-0.dll` in the latest mingw-w64-x86_64-gcc-libs package does appear to be broken. However, this is also likely to have very minimal impact.

I guess I could build my own `git` without `libssp`...

**Comment by lazka** [comment]
hm, It's the first build where we used `-fstack-protector-strong` I think.

**Comment by mati865** [comment]
> I did try building the mingw-w64-x86_64-gcc-libs-12.2.0-5 package myself and I noticed there were some warnings about infinite recursion in the build of `libssp-0.dll`. Could this be the cause?

Can you share logs?

> hm, It's the first build where we used `-fstack-protector-strong` I think.

AFAIK it should not affect libssp itself. Maybe it got messed up by this flag?

> Switching to `libssp-0.dll` from the previous package appears to work fine.

By previous you mean `mingw-w64-x86_64-gcc-libs-12.2.0-4`?

**Comment by jeremyd2019** [comment]
https://github.com/msys2/MINGW-packages/pull/13810#issuecomment-1299854860 but if libssp dll is really broken it should probably be fixed anyway.  @dscho have you seen any issues like this, or is it a red herring?

**Comment by lazka** [comment]
Reproducer:

```console
$ pacman -U https://repo.msys2.org/mingw/mingw64/mingw-w64-x86_64-flac-1.3.4-2-any.pkg.tar.zst
$ flac
*** stack smashing detected ***:  terminated
```

**Comment by dscho** [comment]
Yes, I have seen this, and fixed it in Git for Windows.

About this report: I _do not_ appreciate how blatantly some people disregard the stern warning at the top of https://github.com/git-for-windows/git/wiki/Install-inside-MSYS2-proper.

Anybody is free to do as they wish, but if they use an unsupported setup, they have no right to burden anybody else with fixing their problems with said setup. That's a self-inflicted problem and it is outright unfair to expect others to help solve it.

**Comment by jeremyd2019** [comment]
I'd agree except that now we see it happens to all existing binaries built with stack protection, so there is probably a real issue in how libssp-0 was built (it would be pretty stupid, but maybe it was built with stack protection itself so it is generating code that calls itself?)

**Comment by nga888** [comment]
> >I did try building the mingw-w64-x86_64-gcc-libs-12.2.0-5 package myself and I noticed there were some warnings about infinite recursion in the build of libssp-0.dll. Could this be the cause?
>
> Can you share logs?

Sorry, but I don't have that build around any more and forgot to capture the warnings.

> About this report: I _do not_ appreciate how blatantly some people disregard the stern warning at the top of https://github.com/git-for-windows/git/wiki/Install-inside-MSYS2-proper.

I am fully aware of this warning and did think twice about reporting this issue. However, as I suspected that the `libssp-0.dll` being released is likely broken, I thought it would be appropriate to report even though the use case that revealed the issue is not supported.

**Comment by DavyLandman** [comment]
> I am fully aware of this warning and did think twice about reporting this issue. However, as I suspected that the libssp-0.dll being released is likely broken, I thought it would be appropriate to report even though the use case that revealed the issue is not supported.

I would also interpreted it this way. Like birds in a coal mine, early notifications if something was going wrong.

**Comment by lazka** [comment]
> > hm, It's the first build where we used `-fstack-protector-strong` I think.
> 
> AFAIK it should not affect libssp itself. Maybe it got messed up by this flag?

yeah:

```
L_EXPORT -DPIC -o .libs/strncpy-chk.o
                                                  infinite                                               recursion detected [ In function '-Winfinite-recursion__memmove_chk]
   44 | ':
__memcpy_chk../../../gcc-12.2.0/libssp/memmove-chk.c:45:1: (void *__restrict__ dest, const void *__restrict__ src,
      |  ^~~~~~~~~~~~warning:
                      infinite recursion detected [In file included from ../../../gcc-12.2.0/libssp/memcpy-chk.c:38../../../gcc-12.2.0/libssp/gets-chk.c:48-Winfinite-recursion:
In function '../../../gcc-12.2.0/libssp/memset-chk.c:              gets__memset_chk',
    inlined from '':
__gets_chk../../../gcc-12.2.0/libssp/memset-chk.c:44:1:     :
```

I'm going to disable stack protector for libssp.

**Comment by nga888** [comment]
Glad to see that the issue has been found.

I'd also like to add that the effort that goes into maintaining MSYS2, MinGW and GFW is very much appreciated.

> About this report: I _do not_ appreciate how blatantly some people disregard the stern warning at the top of https://github.com/git-for-windows/git/wiki/Install-inside-MSYS2-proper.
> 
> Anybody is free to do as they wish, but if they use an unsupported setup, they have no right to burden anybody else with fixing their problems with said setup. That's a self-inflicted problem and it is outright unfair to expect others to help solve it.

However, comments like the above do somewhat concern me because it doesn't really help to encourage people to contribute towards these projects.

**Comment by revelator** [comment]
great to hear that it has been solved :) and i agree we have to be a little more carefull with how we react to something that might not be from packages we maintain, sometimes it "can" come back and bite us later on so maybe a little less strict warning ?.

**Comment by dscho** [comment]
> comments like the above do somewhat concern me because it doesn't really help to encourage people to contribute towards these projects.

@nga888 @revelator you have no idea how discouraging _your_ comments are. It really makes me feel very much devalued. Like, you are fine to dump all the work on me, but when I point out that I feel totally taken advantage of, you're saying that I do something wrong? Not cool.

**Comment by revelator** [comment]
ill help as far as my skills go :) im not an ace programmer but i learned a few tricks over the years so feel free to ask me if stuck on something i might have an idea or atleast i can give a different perspective, we are all in this together and im allways willing to lend a hand.

**Comment by nga888** [comment]
> @nga888 @revelator you have no idea how discouraging _your_ comments are. It really makes me feel very much devalued. Like, you are fine to dump all the work on me, but when I point out that I feel totally taken advantage of, you're saying that I do something wrong? Not cool.

I apologise if my comment has made you feel devalued that was not the intention and I certainly didn't want to "dump all the work" on you.

As I said above the effort of all the contributors to these projects is very much appreciated.

I was just trying to say that comments like: "they have no right to burden anybody else with fixing their problems with said setup. That's a self-inflicted problem and it is outright unfair to expect others to help solve it." is not necessarily very welcoming. Particularly, as in this case this is an actual issue with the `libssp-0.dll` that has been released.

I usually locally fix or workaround issues because of my unsupported setup and I have even contributed a fix to GFW in the past. However, in this case because it looked like an actual breakage in the MinGW package, I decided to report it.

Once again, I apologise for any bad feelings caused.

**Comment by jeremyd2019** [comment]
Sorry to have kind of started this - I just wanted to know if this was something dscho had seen in the course of the official git for windows.  But then lazka posted another reproducer so the fact that it was seen in the 'unsupported' git for windows configuration was less relevant.

**Comment by lazka** [comment]
The fixed package (#13836) is now in the repo.
